Dashboard: Mix initials and custom gravatars in recent viewers list

mmandrus opened the pull request:
**What this PR does / why we need it**:
This PR augments the enterprise "recent viewers" feature when looking at dashboards. 

Previously, we displayed either **only** initials or **only** gravatars, depending on the disable_gravatar flag. This change allows for mixing of custom avatars (i.e. set by the user) and initials for "default" avatars (i.e. the retro-style fallback avatars). Because default avatars are random, this change allows users to more easily understand who is viewing dashboards at a glance.

![Screenshot_2022-04-01 11 59 52_zEwd0j](https://user-images.githubusercontent.com/41969079/161304474-6bc874bc-92cb-4f6d-8e79-52d6564aa5cd.png)

**Which issue(s) this PR fixes**: 
grafana-enterprise/544

**Other**
- This has a corresponding PR on the grafana-enterprise side
- Thank you @pianohacker for helping decide the implementation approach

mmandrus committed:
Use Wiring to initialize Avatar Cache Server

Create AvatarCacheServer Provider function and pass it in as an
argument to HTTPServer. Also convert CacheServer to a singleton
so that we keep all cached Avatar info in one place for easier access

mmandrus committed:
Refactor avatar cache server and add 'isCustom' check

Avatar cache server needs to perform two similar fetches
back-to-back; break up functions to allow for easy reuse.
Then add handling to see if a user has a custom avatar.

mmandrus committed:
Add additional accessors so that /recents api can easily use the cache

mmandrus committed:
merge main -- resolve conflict in wire.go

mmandrus committed:
Minor mods to avatar server to facilitiate unit testing

mmandrus committed:
add unit tests for avatar fetching

mmandrus committed:
Merge remote-tracking branch 'origin/main' into mmandrus/recents-avatar-initials-mixing

mmandrus committed:
add error handling in case we somehow fetch gravatars while they are disabled

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=47212) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=47212) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=47212) it.</sub>

mmandrus commented:
Has a failing front end unit test due to the backend change, update coming soon

mmandrus committed:
linting: read error return value in unit test

leandro-deveikis reviewed (approved):
LGTM!

evictorero commented on pkg/api/avatar/avatar.go:
```diff
@@ -214,11 +259,48 @@ var client = &http.Client{
+
+// Verifies the Gravatar response code was 200, then stores the image byte slice
+func getGravatarHandler(av *Avatar, resp *http.Response) error {
+	if resp.StatusCode != 200 {
```
```suggestion
	if resp.StatusCode != http.StatusOK {
```

evictorero commented on pkg/api/avatar/avatar.go:
```diff
@@ -214,11 +259,48 @@ var client = &http.Client{
+
+// Uses the d=404 fallback to see if the gravatar we got back is custom
+func checkIsCustomHandler(av *Avatar, resp *http.Response) error {
+	av.isCustom = resp.StatusCode != 404
```
```suggestion
	av.isCustom = resp.StatusCode != http.StatusNotFound
```

evictorero commented on pkg/api/avatar/avatar.go:
```diff
@@ -227,9 +309,10 @@ func (a *thunderTask) fetch() error {
 	resp, err := client.Do(req)
 	if err != nil {
-		a.Avatar.notFound = true
+		av.HandleErrorResponse()
```
If an error occurs we return an error. Is it really needed to set the `Avatar` values before returning an error?

mmandrus replied:
In this case, yes. The flags we set on the avatar tell the endpoint to return the default "not found" image. We could use default values instead, but I think setting them explicitly every time is safer.

evictorero reviewed (commented):
Great work @mmandrus ! Love this feature, makes a lot of sense. I left some comments and nits.

mmandrus commented on pkg/api/avatar/avatar.go:
```diff
@@ -61,31 +70,67 @@ func (a *Avatar) Encode(wr io.Writer) error {
 	case <-time.After(time.Second * 3):
 		err = fmt.Errorf("get gravatar image %s timeout", a.hash)
-	case err = <-thunder.GoFetch(gravatarSource+a.hash+"?"+a.reqParams, a):
+	case err = <-thunder.GoFetch(baseUrl, a):
```
Yes, the params get appended as needed in the thunderTask.fetch() function

mmandrus commented on pkg/api/avatar/avatar.go:
```diff
@@ -61,31 +70,67 @@ func (a *Avatar) Encode(wr io.Writer) error {
 
-func (a *Avatar) Update() (err error) {
+func (a *Avatar) update(baseUrl string) (err error) {
+	baseUrl += a.hash + "?"
```
okay, will assign to a new variable

mmandrus committed:
Use http package status codes

Co-authored-by: Ezequiel Victorero <evictorero@gmail.com>

mmandrus committed:
Use http package status codes

Co-authored-by: Ezequiel Victorero <evictorero@gmail.com>

mmandrus committed:
Use http package status codes

Co-authored-by: Ezequiel Victorero <evictorero@gmail.com>

mmandrus committed:
Incorporate suggestions from PR
-avoid mutating arguments
-change error handler function to private and make name more descriptive

evictorero reviewed (approved):
👌

jalevin commented on pkg/api/dtos/models.go:
```diff
@@ -78,11 +78,20 @@ func GetGravatarUrl(text string) string {
 		mlog.Warn("Failed to hash text", "err", err)
 	}
-	return fmt.Sprintf(setting.AppSubUrl+"/avatar/%x", hasher.Sum(nil))
+	return hasher.Sum(nil), true
```
Would it be better to return an empty byte slice rather than performing sum operation on nil?
